stdlib/src/utils/string_slice.mojo

```diff
@@ -26,6 +26,129 @@ alias StaticString = StringSlice[ImmutableStaticLifetime]
 """An immutable static string slice."""
 
 
+fn is_valid_utf8[
+    width: Int = 8
+](data: UnsafePointer[UInt8], length: Int) -> Bool:
+    """Verify that the bytes are valid UTF-8.
+
+    Parameters:
+        width: The width of the SIMD vector to use for validation when length
+            fits. The rest is verified per byte.
+
+    Args:
+        data: The pointer to the data.
+        length: The length of the items pointed to.
+
+    Returns:
+        Whether the data is valid UTF-8.
+
+    Constraints:
+        - width must be >= 4.
+
+    #### UTF-8 coding format
+    [Table 3-7 page 94](http://www.unicode.org/versions/Unicode6.0.0/ch03.pdf).
+    Well-Formed UTF-8 Byte Sequences
+
+    Code Points        | First Byte | Second Byte | Third Byte | Fourth Byte |
+    :----------        | :--------- | :---------- | :--------- | :---------- |
+    U+0000..U+007F     | 00..7F     |             |            |             |
+    U+0080..U+07FF     | C2..DF     | 80..BF      |            |             |
+    U+0800..U+0FFF     | E0         | ***A0***..BF| 80..BF     |             |
+    U+1000..U+CFFF     | E1..EC     | 80..BF      | 80..BF     |             |
+    U+D000..U+D7FF     | ED         | 80..***9F***| 80..BF     |             |
+    U+E000..U+FFFF     | EE..EF     | 80..BF      | 80..BF     |             |
+    U+10000..U+3FFFF   | F0         | ***90***..BF| 80..BF     | 80..BF      |
+    U+40000..U+FFFFF   | F1..F3     | 80..BF      | 80..BF     | 80..BF      |
+    U+100000..U+10FFFF | F4         | 80..***8F***| 80..BF     | 80..BF      |
+    .
+    """
+
+    constrained[width >= 4, "width must be >= 4."]()
+    var ptr = DTypePointer(data)
+    var iter_len = length
+    # TODO: implement a faster algorithm like https://github.com/cyb70289/utf8
+    # and benchmark the difference.
+
+    fn invalid_special(b0: UInt8, b1: UInt8) -> Bool:
+        """Special unicode ranges."""
+        if b0 == 0xE0 and not (UInt8(0xA0) <= b1 <= UInt8(0xBF)):
+            return True
+        elif b0 == 0xED and not (UInt8(0x80) <= b1 <= UInt8(0x9F)):
+            return True
+        elif b0 == 0xF0 and not (UInt8(0x90) <= b1 <= UInt8(0xBF)):
+            return True
+        elif b0 == 0xF4 and not (UInt8(0x80) <= b1 <= UInt8(0x8F)):
+            return True
+        return False
+
+    var idx = 0
```

Comment:
Could you please provide some comments explaining this algorithm? Also, consider better variable names, e.g. `n2_m` below is not very meaningful.